File: solutions/typescript/problems/[0138_Medium] Copy List with Random Pointer.ts
class RanNode {
    val: number;
    next: RanNode | null;
    random: RanNode | null;
    constructor(val?: number, next?: RanNode, random?: RanNode) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
        this.random = random === undefined ? null : random;
    }
}

function copyRandomList(head: RanNode | null): RanNode | null {
    let curr: RanNode | null = head;
    // First pass - create duplicate nodes
    while (curr) {
        const next: RanNode | null = curr.next;
        curr.next = new RanNode(curr.val, next || undefined, curr.random || undefined);
        curr = next;
    }
    // Second pass - update random pointers of the duplicate nodes
    curr = head?.next || null;
    while (curr) {
        curr.random = curr.random?.next || null;
        curr = curr.next?.next || null;
    }
    // Third pass - remove original nodes
    const newHead = head?.next || null;
    curr = head;
    while (curr) {
        const next: RanNode | null = curr.next;
        curr.next = next?.next || null;
        curr = next;
    }
    return newHead;
}

File: solutions/c++/problems/[0138_Medium] Copy List with Random Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node *copyRandomList(Node *head) {
        if (!head) return head;
        auto curr = head;
        while (curr) {
            auto next = curr->next;
            auto copyNode = new Node(curr->val);
            copyNode->next = next;
            copyNode->random = curr->random;
            curr->next = copyNode;
            curr = next;
        }
        curr = head;
        while (curr) {
            auto copiedNode = curr->next;
            if (copiedNode->random) copiedNode->random = copiedNode->random->next;
            curr = copiedNode->next;
        }
        curr = head;
        auto newHead = head->next;
        while (curr) {
            auto copiedNode = curr->next;
            auto next = copiedNode->next;
            curr->next = next;
            if (next) {
                copiedNode->next = next->next;
            }
            curr = next;
        }
        return newHead;
    }
};